q2_metadata/normalization/_norm_rules.py

```diff
@@ -0,0 +1,322 @@
+                        self.rules['allowed'][rule] = rule_value
+                    else:
+                        errorsCollection(self.variable, rule, rule_value)
+
+            elif rule == 'ontology':
+                if isinstance(rule_value, str):
+                    if rule_value in ['Gazetteer']:
+                        self.rules['lookups'][rule] = rule_value
+
+            elif rule == 'remap':
+                if isinstance(rule, dict):
+                    str_values = [x for x,y in rule_value.items() if isinstance(y, str)]
+                    if len(str_values) == len(rule_value.keys()):
+                        self.rules['edits'][rule] = rule_value
+                    else:
+                        errorsCollection(self.variable, rule, rule_value)
+
+            elif rule == 'validation':
+                possible_validations = {
+                    'force_to_blank_if': {'is_null'}
+                }
+                if isinstance(rule_value, dict):
+                    rule_value_set = set(rule_value)
+                    if len(rule_value_set ^ (rule_value_set & set(possible_validations))):
+                        errorsCollection(self.variable, rule, rule_value)
+                        # raise and stop ? or keep checking?
+                    else:
+                        for rule_2, rule_value_2 in rule_value['force_to_blank_if'].items():
+                            if isinstance(rule_value_2, list):
+                                possible_validations_2 = possible_validations['force_to_blank_if']
+                                if len(set(rule_2) ^ (set(rule_2) & possible_validations_2)):
+                                    errorsCollection(self.variable, rule, rule_value)
+                                else:
+                                    self.rules['edits'][rule] = rule_value
+                            else:
+                                errorsCollection(self.variable, rule, rule_value)
+                else:
+                    errorsCollection(self.variable, rule, rule_value)
+
+            elif rule == 'normalization':
+                if isinstance(rule_value, dict):
+                    possible_normalizations = {'maximum', 'mininum', 'gated_value'}
+                    rule_value_set = set(rule_value)
+                    if len(rule_value_set ^ (rule_value_set & possible_normalizations)):
+                        errorsCollection(self.variable, rule, rule_value)
+                    else:
+                        self.rules['edits'][rule] = rule_value
+
+            elif rule == 'blank':
+                if isinstance(rule_value, str):
+                    self.rules['allowed'][rule] = rule_value
+
+            elif rule == 'missing':
+                if isinstance(rule_value, str):
+                    self.rules['allowed'][rule] = rule_value
+
+            elif rule == 'format':
+                if isinstance(rule_value, str):
+                    self.rules['format'][rule] = rule_value
```

Comment:
Can you add an "else" clause and raise an exception? This means that the rule type is not known.

Reply:
*Can you add an "else" clause and raise an exception?*
Okay

* This means that the rule type is not known.*
The thing is that we have to check that that the user-editable rules file is passing each rule in its accepted format. Please let me know if there is a preferred way to do this?

Reply:
I plan on making a set of user-defined exceptions classes for the rules-format checking (as in the [official doc](https://docs.python.org/3/tutorial/errors.html#user-defined-exceptions)), that would be located in another module (`_norm_format_exceptions.py`) that would collect the errors. These errors would be then redirected to the `ErrorsWarnings()` instance.
Do you think this a good way to go?

Reply:
For this, I have made substantial change:

Inside the Rules() class (instantiated for each variable), instead of having the method `_process_rule()` which created what I think is cyclomatic complexity, I have separated the parsing of the yaml file (static method `parse_rule()`) and the checking of the rules formatting (method `_check_variable_rules()`).

This new function - `_check_variable_rules()` - simply starts a loop, which - for each variable' iteration - runs a `check_rule()` function defined in a separate module (`q2_metadata/normalization/_norm_check_rules.py`). For each possible rule of the variable that's iterated over, this `check_rule()` function either:
- fill the data structure that's an essential attribute of the Rules() class: `self.rules` and that is contained into the variable `rules` in the main script (`q2_metadata/_normalize()`)
- throw an error, that is one of different possible user-defined exceptions from the new module `q2_metadata/normalization/_norm_errors.py`.

Important: each of the per-rule functions in `_norm_check_rules.py` and the user-defined exception in `_norm_errors.py` are unittested, in `q2_metadata/tests/normalization/test_norm_errors.py` and `q2_metadata/tests/normalization/test_errors_classes.py`, respectively.

Reply:
I have further modified the code in `q2_metadata/normalization/_norm_check_rules.py ` so that the __init__ of:

- `RulesCollection()` only get an empty `self.variables_rules` and it is in the main script that this attribute is populated (see `rules.parse_variables_rules(list(variables_rules_files))`)
--> the keys of `self.variables_rules` are then used compared with the metadata variables and only those in common (the "focus") are checked. This check is now also made from the main script (see `rules.check_variables_rules(list(focus))`).

- `Rules()` only gets - for each rule - the raw parses from the yaml and the empty, homogenised rules data structure `self.rules`. For now, the method the does the cleaning `normalize()` is empty and it should be into another branch.

I have added a lot of unittests (along with test, dummy rules files), see below (the test_*.py are at the bottom):
```
.
├── __init__.py
├── normalization
│   ├── __init__.py
│   ├── metadata
│   │   ├── franck_clean.tsv
│   ├── rules
│   │   ├── rules_collection_errors
│   │   │   ├── empty
│   │   │   ├── no_yml
│   │   │   │   └── touch.txt
│   │   │   └── ok
│   │   │       ├── A.yml
│   │   │       └── B.yml
│   │   └── rules_errors
│   │       ├── not_yml
│   │       │   ├── A.yml.gz
│   │       │   └── B.yml
│   │       ├── ok
│   │       │   ├── all_ok.yml
│   │       │   └── simple.yml
│   │       └── raisers
│   │           ├── BlankError
│   │           │   ├── BlankError_0.yml
│   │           │   ├── BlankError_1.yml
│   │           │   └── BlankError_2.yml
│   │           ├── ExpectedError
│   │           │   ├── ExpectedError_0.yml
│   │           │   ├── ExpectedError_1.yml
│   │           │   └── ExpectedError_2.yml
│   │           ├── FormatError
│   │           │   ├── FormatError_0.yml
│   │           │   ├── FormatError_1.yml
│   │           │   └── FormatError_2.yml
│   │           ├── MissingError
│   │           │   ├── MissingError_0.yml
│   │           │   ├── MissingError_1.yml
│   │           │   └── MissingError_2.yml
│   │           ├── NormalizationError
│   │           │   ├── NormalizationError_0.yml
│   │           │   ├── NormalizationError_1.yml
│   │           │   └── NormalizationError_2.yml
│   │           ├── OntologyError
│   │           │   ├── OntologyError_0.yml
│   │           │   ├── OntologyError_1.yml
│   │           │   └── OntologyError_2.yml
│   │           ├── RemapError
│   │           │   ├── RemapError_0.yml
│   │           │   ├── RemapError_1.yml
│   │           │   └── RemapError_2.yml
│   │           └── ValidationError
│   │               ├── ValidationError_0.yml
│   │               ├── ValidationError_1.yml
│   │               ├── ValidationError_2.yml
│   │               └── ValidationError_3.yml
│   ├── test_errors_classes.py
│   ├── test_norm_errors.py
│   ├── test_norm_rules.py
│   └── test_utils.py
```